tools/shared.py

```diff
@@ -2612,6 +2609,51 @@ def minify_wasm_imports_and_exports(js_file, wasm_file, minify_whitespace, minif
     extra_info = {'mapping': mapping}
     return Building.acorn_optimizer(js_file, passes, extra_info=json.dumps(extra_info))
 
+  @staticmethod
+  def wasm2js(js_file, wasm_file, opt_level, minify_whitespace, use_closure_compiler, debug_info):
+    logger.debug('wasm2js')
+    cmd = [os.path.join(Building.get_binaryen_bin(), 'wasm2js'), '--emscripten', wasm_file]
+    if opt_level > 0:
+      cmd += ['-O']
+    cmd += Building.get_binaryen_feature_flags()
+    wasm2js_js = run_process(cmd, stdout=PIPE).stdout
+    if opt_level >= 2:
+      temp = configuration.get_temp_files().get('.js').name
+      with open(temp, 'a') as f:
+        f.write(wasm2js_js)
+      # if closure is 1, minify the code in a non-advanced way here (we don't validate
+      # as asm.js, and so can use a stock minifier like closure). if we are in closure 2
+      # mode, then we'll minify it all together later on
+      if use_closure_compiler == 1:
+        temp = Building.closure_compiler(temp,
+                                         pretty=not minify_whitespace,
+                                         advanced=False)
+      elif minify_whitespace:
+        temp = Building.js_optimizer_no_asmjs(temp, ['minifyWhitespace'])
+      with open(temp) as f:
+        wasm2js_js = f.read()
+      # closure may leave a trailing `;`, which would be invalid given where we
```

Comment:
given where we what?

Reply:
Oops, yeah, the comment got cut off somehow. I'll fix. The point is that we place this code inside parens, and it's not valid to have a `;` there.